### PR TITLE
Snapshot listener source from cache

### PR DESCRIPTION
**Pull request by milaGGL**
ported from https://github.com/firebase/firebase-js-sdk/pull/7982

### CODE REVIEW

File: Firestore/core/src/core/sync_engine.cc [hunk 1]
```diff
@@ -176,11 +185,50 @@ void SyncEngine::StopListening(const Query& query) {
...
+  TargetId target_id = query_view->target_id();
+  auto& queries = queries_by_target_[target_id];
+  queries.erase(std::remove(queries.begin(), queries.end(), query),
+                queries.end());
+
+  if (queries.empty()) {
+    remote_store_->StopListening(target_id);
+  }
+}
+// void SyncEngine::StopListeningAndReleaseTarget(const Query& query, bool
```
Comment (wu-hui) [inline Firestore/core/src/core/sync_engine.cc:210]: delete this?

File: Firestore/core/src/core/sync_engine.cc [hunk 2]
```diff
@@ -176,11 +185,50 @@ void SyncEngine::StopListening(const Query& query) {
...
 
+void SyncEngine::StopListeningToRemoteStore(const Query& query) {
+  AssertCallbackExists("stopListeningToRemoteStore");
+
+  auto query_view = query_views_by_query_[query];
+  HARD_ASSERT(query_view, "Trying to stop listening to a query not found");
+
+  TargetId target_id = query_view->target_id();
+  auto& queries = queries_by_target_[target_id];
+  queries.erase(std::remove(queries.begin(), queries.end(), query),
```
Comment (wu-hui) [inline Firestore/core/src/core/sync_engine.cc:203]: std::remove is a c++17/20 feature, which we cannot use.

File: Firestore/core/test/unit/core/event_manager_test.cc [hunk 3]
```diff
@@ -73,14 +75,14 @@ TEST(EventManagerTest, HandlesManyListnersPerQuery) {
   EXPECT_CALL(mock_event_source, SetCallback(_));
   EventManager event_manager(&mock_event_source);
 
-  EXPECT_CALL(mock_event_source, Listen(query));
+  EXPECT_CALL(mock_event_source, Listen(query, true));
```
Comment (wu-hui) [inline Firestore/core/test/unit/core/event_manager_test.cc:78]: There should be some tests to cover when the second parameter is false.

File: Firestore/Example/Tests/Integration/API/FIRSnapshotListenerSourceTests.mm [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+
+#import "Firestore/Example/Tests/Util/FSTEventAccumulator.h"
+#import "Firestore/Example/Tests/Util/FSTHelpers.h"
+#import "Firestore/Example/Tests/Util/FSTIntegrationTestCase.h"
+#import "Firestore/Example/Tests/Util/FSTTestingHooks.h"
+
+#import "Firestore/Source/API/FIRAggregateQuerySnapshot+Internal.h"
+#import "Firestore/Source/API/FIRQuery+Internal.h"
+
+@interface FIRSnapshotListenerSourceTests : FSTIntegrationTestCase
```
Comment (wu-hui) [inline Firestore/Example/Tests/Integration/API/FIRSnapshotListenerSourceTests.mm:29]: It's my fault to not tell you this earlier, but we'd like new integration tests to be in swift going forward.

**Review by wu-hui — COMMENTED** [review]
Sorry, i have to ask you to rewrite your integration tests in swfit.

File: Firestore/core/src/core/sync_engine.cc (hunk 1 above)
Comment (milaGGL) [reply Firestore/core/src/core/sync_engine.cc:210]: oops. forgot to remove it.

File: Firestore/Example/Tests/Integration/API/FIRSnapshotListenerSourceTests.mm (hunk 4 above)
Comment (milaGGL) [reply Firestore/Example/Tests/Integration/API/FIRSnapshotListenerSourceTests.mm]: objective c tests are moved to swift

File: Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+typedef NS_ENUM(NSUInteger, FIRListenSource) {
+  /**
+   * The default behavior. The listener attempts to return initial snapshot from cache and retrieve
+   * up-to-date snapshots from the Firestore server. Snapshot events will be triggered on local
+   * mutations and server side updates.
+   */
+  FIRListenSourceDefault,
+  /**
+   * The listener retrieves data and listens to updates from the local Firestore cache only. If the
+   * cache is empty, an empty snapshot will be returned. Snapshot events will be triggered on cache
```
Comment (wu-hui) [inline Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h:33]: It is not clear what "Snapshot events" are to the users.

I think we can reword to something like:

The listener retrieves data and listens to updates from the local Firestore cache without attempting to send the query to the server. If some documents gets updated as a result from other queries, they will be picked up by listeners using `Cache`.

**Review by wu-hui — APPROVED** [review]
Congrats!!

File: Firestore/Source/API/FIRSnapshotListenOptions.mm [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+#include <cstdint>
+#include <string>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation FIRSnapshotListenOptions
+
+// private constructor method
+- (instancetype)initPrivateWithSource:(FIRListenSource)source
```
[the remark this answers is not in the document]
Comment (milaGGL) [reply Firestore/Source/API/FIRSnapshotListenOptions.mm:29]: nope, it have a comment, so it should be enough. I will remove the "private"

File: Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h (hunk 5 above)
Comment (milaGGL) [reply Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h:33]: sounds good

File: Firestore/Source/API/FIRSnapshotListenOptions.mm (hunk 6 above)
[the remark this answers is not in the document]
Comment (milaGGL) [reply Firestore/Source/API/FIRSnapshotListenOptions.mm:29]: wait, it is quite common in our sdk to name the private constructor "initPrivate": 
Screenshot 2024-03-05 at 3 47 40 PM
thats why I added the "private" at the first place. 
I will remove the comment as it is implied by the function name already

File: Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * The source that snapshot listener retrieves data from.
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h:22]: "The source the snapshot listener...." (replace "that")

File: Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  /**
+   * The default behavior. The listener attempts to return initial snapshot from cache and retrieve
+   * up-to-date snapshots from the Firestore server. Snapshot events will be triggered on local
+   * mutations and server side updates.
+   */
+  FIRListenSourceDefault,
+  /**
+   * The listener retrieves data and listens to updates from the local Firestore cache without
+   * attempting to send the query to the server. If some documents gets updated as a result from
+   * other queries, they will be picked up by listeners using Cache.
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h:34]: "...listeners using the cache."

File: Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * The source that snapshot listener retrieves data from.
+ */
+typedef NS_ENUM(NSUInteger, FIRListenSource) {
+  /**
+   * The default behavior. The listener attempts to return initial snapshot from cache and retrieve
+   * up-to-date snapshots from the Firestore server. Snapshot events will be triggered on local
+   * mutations and server side updates.
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h:28]: "server-side"

File: Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+ * Creates and returns a new `SnapshotListenOptions` object with all properties initialized to their
+ * default values.
+ *
+ * @return The created `SnapshotListenOptions` object.
+ */
+- (instancetype)init NS_DESIGNATED_INITIALIZER;
+
+/**
+ * Creates and returns a new `SnapshotListenOptions` object with with all properties of the current
+ * `SnapshotListenOptions` object and the new configuration of whether metadata-only changes should
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h:65]: "...with all properties of the current `SnapshotListenOptions` object plus the new property specifying whether  metadata-only changes should trigger snapshot events."

File: Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+ * trigger snapshot events.
+ *
+ * @return The created `SnapshotListenOptions` object.
+ */
+- (FIRSnapshotListenOptions *)optionsWithIncludeMetadataChanges:(BOOL)includeMetadataChanges;
+
+/**
+ * Creates and returns a new `SnapshotListenOptions` object with with all properties of the current
+ * `SnapshotListenOptions` object and the new configuration of the source that snapshot listener
```
Comment (markarndt) [inline Firestore/Source/Public/FirebaseFirestore/FIRSnapshotListenOptions.h:74]: "...with all properties of the current `SnapshotListenOptions` object plus the new property specifying the source that the..."

File: Firestore/core/src/api/listen_source.h [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+#ifndef FIRESTORE_CORE_SRC_API_LISTEN_SOURCE_H_
+#define FIRESTORE_CORE_SRC_API_LISTEN_SOURCE_H_
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+/**
+ * An enum that configures the source of snapshot listeners listens to. By
```
Comment (markarndt) [inline Firestore/core/src/api/listen_source.h:25]: "...that configures the snapshot listener data source."

"Using this enum, specify whether snapshot events are triggered by..."

**Review by markarndt — APPROVED** [review]

**Pull request merged**